derivationStrict

layus opened the pull request:
Implement derivationStrict primOp

Closes #364

sorki commented on src/Nix/Effects.hs:
```diff
@@ -226,36 +242,58 @@ print = putStrLn . show
+    path <- parseStoreResult "addTextToStore" res
+    case path of
+      Left error -> return $ Left error
+      Right path -> return $ Right $ StorePath $ T.unpack $ T.decodeUtf8 $ Store.storePathToRawFilePath path
```
`storepath` branch now has `storePathToText` iff needed but it would be better to switch to `StorePath` type of `hnix-store-core`.

layus replied:
Yep, there is definitely something to do about this. expipiplus1 is also quite bothered with that (https://github.com/haskell-nix/hnix/pull/554/files#r538011509).

I guess https://github.com/haskell-nix/hnix/issues/472 is the right tracking issue.

sorki commented on src/Nix/Effects.hs:
```diff
@@ -226,36 +242,58 @@ print = putStrLn . show
+    Right pathName -> do
+      -- todo: pass the hash explicitly
+      -- TODO: Handle the filter
+      res <- Store.runStore $ Store.addToStore pathName path recursive (Proxy :: Proxy Store.SHA256) (const False) repair
```
This also changed type to omit proxy so you can just do `Store.addToStore @'Store.SHA256 pathName ..`

layus committed:
Cache hashDerivationModulo results ourselves

Ideally, we would have this cache inside the (h)nix-store, and persist
the store connection for the whole session.  Consider this a proof of
concept that may last.

layus committed:
Implement derivationStrict primOp

layus committed:
Principled nvpath -> nvstring coercion

layus committed:
Allow custom computations inside string contexts

expipiplus1 commented on default.nix:
```diff
@@ -223,6 +222,17 @@ let
+      # 2020-12-07 hnix-store-remote fails when trying to connect to a real hnix daemon.
+      # probably due to nix sandbox restrictions.
+      hnix-store-remote = pkgs.haskell.lib.dontCheck super.hnix-store-remote;
+
```
Just a side note, have you made PRs for these to NixOS/nixpkgs?

layus replied:
I did not, but now it's done :-)

- cryptohash-sha512:  https://github.com/NixOS/nixpkgs/pull/106333
- hnix-store-remote: https://github.com/haskell-nix/hnix-store/issues/80

expipiplus1 commented on src/Nix/Effects.hs:
```diff
@@ -226,36 +235,54 @@ print = putStrLn . show
+    addToStore a b c d = lift $ addToStore a b c d
+
+    -- | Add a nar (action) to the store
+    -- addToStore' :: StorePathName -> IO Nar -> RecursiveFlag -> RepairFlag -> m (Either ErrorCall StorePath)
```
remove?

expipiplus1 commented on src/Nix/Effects.hs:
```diff
@@ -226,36 +235,54 @@ print = putStrLn . show
+  addToStore name path recursive repair = case Store.makeStorePathName name of
+    Left err -> return $ Left $ ErrorCall $ "String '" ++ show name ++ "' is not a valid path name: " ++ err
+    Right pathName -> do
+      -- TODO: redesign the filter parameter
```
Can add an issue for this?

layus replied:
I have a WIP PR. Not really an issue though: https://github.com/haskell-nix/hnix/pull/755

expipiplus1 commented on src/Nix/Effects.hs:
```diff
@@ -226,36 +235,54 @@ print = putStrLn . show
+      res <- Store.runStore $ Store.addToStore @'Store.SHA256 pathName path recursive (const False) repair
+      parseStoreResult "addToStore" res >>= \case
+        Left err -> return $ Left err
+        Right storePath -> return $ Right $ StorePath $ T.unpack $ T.decodeUtf8 $ Store.storePathToRawFilePath storePath
```
I guess it's no problem for nix store paths, but I always get the heebie-jeebies seeing encoding/decoding near file names.

layus replied:
Yes, maybe make a further issue for it. As long as hnix-store uses ByteStrings and hnix FilePaths, this will remain a pain. That being said, If this could be the only place where the conversion happens, it remains quite clean. At the moment such conversions also happen here and there in Effects/Derivation.hs.

layus replied:
I guess https://github.com/haskell-nix/hnix/issues/472 is the right tracking issue.

expipiplus1 commented on src/Nix/Effects/Basic.hs:
```diff
@@ -235,13 +225,13 @@ findPathM = findPathBy existingPath
 
 defaultImportPath
-  :: (MonadNix e t f m, MonadState (HashMap FilePath NExprLoc) m)
+  :: (MonadNix e t f m, MonadState (HashMap FilePath NExprLoc, b) m)
```
Shame that Haskell/mtl make the `b` necessary here, I guess it's just to make it fit a hole somewhere else.

layus replied:
We could move to a more involved type. Possibly something like `Has ExprCache` and `Has DrvHashCache` but for StateT instead of ReaderT. Looks a lot lensy to me. I did not want to bother generalizing too much because

1. This might move to hnix-store if we end-up weaving the effect stacks
2. There does not seem to be a need for more stateful values
3. Usage of this state is limited to a few fucntions.

expipiplus1 replied:
Oh sure, I was just musing :)

expipiplus1 commented on src/Nix/Reduce.hs:
```diff
@@ -66,19 +67,19 @@ import           System.FilePath
 newtype Reducer m a = Reducer
     { runReducer :: ReaderT (Maybe FilePath, Scopes (Reducer m) NExprLoc)
-                           (StateT (HashMap FilePath NExprLoc) m) a }
+                           (StateT (HashMap FilePath NExprLoc, MS.HashMap Text Text) m) a }
```
Is it possible to give this Text, Text Map a type synonym (or newtype even if it's no trouble) to make it a bit more meaningful?

layus replied:
Again, this is supposed to be a short-lived state addition, but we could certainly do that.

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+  , outputs :: Map Text Text
+  , inputs :: (Set Text, Map Text [Text])
+  , platform :: Text
+  , builder :: Text -- should be typed as a store path
```
Can add an issue for this TODO? (Or implement it I guess)

layus replied:
What we want here is a type that carries the proof that `builder` is a valid path in the store. A `StorePath` does not immediately provide that. A `Store.StorePath` (a `StorePath` from hnix-store) does, but then we mix and match more than we would like to. Will keep as a TODO until we decide on migrating this code to the store, or at least unify the Derivation datatype.

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+  , platform :: Text
+  , builder :: Text -- should be typed as a store path
+  , args :: [ Text ]
+  , env :: Map Text Text
```
Text is probably fine for most of these, but the env one does feel like it should be a bytestring. 

would it be possible to switch these over to bytestring (or short bytestring)? Or better yet newtypes around bytestring. 

Perhaps nix specifies that `getEnv` only works with well-encoded environment vars though, in which case I guess this is fine.

layus replied:
`env` in this context is the set of environment variables available to the builder (`buildCommand`, `cmakeFlags` and the like). I am not sure how these are passed, but as they come from nix files that are parsed as text, having them as text does not cause any loss of precision, right ? Something else to ponder when unifying Dereivation datatype with the store.

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+      let (hashType, hashMode) = case Text.splitOn ":" rht of
+            ["r", ht] -> (ht, Recursive)
+            [ht] ->      (ht, Flat)
+            _ -> undefined -- What ?! -- TODO: Throw a proper error
```
I guess `error "impossible: empty text in parseFixed"` would be slightly better, but no big deal either way

layus replied:
Yes. What I wanted is to make the parser fail with a parse error. But `error` will do for now :-)

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+            _ -> undefined -- What ?! -- TODO: Throw a proper error
+      in case Store.mkNamedDigest hashType hash of
+        Right digest -> (Just digest, hashMode)
+        Left _err -> undefined -- TODO: Raise a proper parse error.
```
Not sure on this one, should probably be at least `error ".."`

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+
+    -- Memoize here, as it may be our last chance in case of readonly stores.
+    drvHash <- Store.encodeBase16 <$> hashDerivationModulo drv'
+    modify (\(a, b) -> (a, MS.insert drvPath drvHash b))
```
This goes for all these modifies, the tuple has added a layer of laziness. Have you checked that this is sufficiently strict to not be a performance problem?

layus replied:
No, I did not and I am not aware of any way to do so. I will provide a quick, strict datatype so as to address this concern and naming as stated above.

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+    let outputsWithContext = Map.mapWithKey (\out path -> principledMakeNixStringWithSingletonContext path (StringContext drvPath (DerivationOutput out))) (outputs drv')
+        drvPathWithContext = principledMakeNixStringWithSingletonContext drvPath (StringContext drvPath AllOutputs)
+        attrSet = M.map nvStr $ M.fromList $ ("drvPath", drvPathWithContext): Map.toList outputsWithContext
+    -- TODO: Add location information for all the entries.
```
can add issue?

layus replied:
Well, I am not too sure what the location should be. I gues the right location is the builtin invocation, which is not very relevant as there is usually only one builtin.derivationStrict, and that is inside builtin.derivation.

Also, I am not sure how to access that location.

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+      AllOutputs ->
+        -- TODO: recursive lookup. See prim_derivationStrict
+        -- XXX: When is this really used ?
+        undefined
```
can change to `error`?

expipiplus1 commented on src/Nix/Effects/Derivation.hs:
```diff
@@ -0,0 +1,395 @@
+      mFixedOutput <- case mHash of
+        Nothing -> return Nothing
+        Just hash -> do
+          when (outputs /= ["out"]) $ lift $ throwError $ ErrorCall $ "Multiple outputs are not supported for fixed-output derivations"
```
I don't know on the nix side of things, but could this work with a single output not named "out"?

layus replied:
At the moment the nix daemon will refuse fixed-outptut derivations that have custom outputs, including a single one not named "out". We just fail earlier.

expipiplus1 reviewed (approved):
Looks great, happy to merge! Any of the nits I mentioned could be done now or have issues created and done at a later time.

It would be good to have some more tests for the various improvements here, but that could also be done later.

sorki reviewed (approved):
LGTM!

I would love to see this use `System.Nix.Derivation` based on `nix-derivation` but since it doesn't map 1:1 and possibly needs another wrapper we can poke that later.

Nice work!

layus committed:
Add references, docs and minor improvements

default.nix: cosmetic: fx align

layus reviewed (commented):
Added one commits to address all the received comments.

I tried to make the eval state typed, but failed halfway. I am not even sure what I am manipulating. It looks like all the types align magically, but that there might be no centralized place to define them. It creates circular inclusions.
Looks like a nontrivial refactor is needed for that. See https://github.com/layus/hnix/commit/63ab545675e00eb58b867a405037f95fc59a6c76 for my experiment.

Anton-Latukha committed:
cabal.project: override the cryptohash-sha512

This should be temporary.

bqv commented on hnix.cabal:
```diff
@@ -401,8 +402,9 @@ library
     , hashing >= 0.1.0 && < 0.2
-    , hnix-store-core >= 0.1.0 && < 0.3
-    , http-client >= 0.5.14 && < 0.6 || >= 0.6.4 && < 0.8
+    , hnix-store-core >= 0.3.0 && < 0.4
```
@layus if you wouldn't mind updating these bounds for the latest release, perhaps @expipiplus1 would be happy to merge this?

Anton-Latukha committed:
use & support recent hnix-store `master` == 0.4

M  default.nix
M  hnix.cabal
M  src/Nix/Effects/Derivation.hs

default.nix: fx overlay call

layus committed:
Update chroot store syntax to new nix standards

layus committed:
install required data files with the binary
